### PR TITLE
Use TORCH_EXTENSION_NAME macro to avoid mismatched module/extension name

### PR DESCRIPTION
**Pull request by goldsborough**
For C++ extensions, currently, the module name (defined via  `PYBIND11_MODULE`) must match the extension name (defined in `setup.py` or `torch.utils.cpp_extension.load`). This can cause problems when users mistype one of them.

To fix this, I first wrote a check that would look for the `PYBIND11_MODULE` declaration in any of the provided source files, and compare the module name declared there with the extension name. This scaled fine and would work. You can find this approach in the first commit of this PR.

However, @zdevito then had a better idea of just defining a macro via a compiler flag that would contain the extension name, that users could then use. I named this `TORCH_EXTENSION_NAME` and it works very nicely.

The other changes are just fixes to the C++ extension tests.

@zdevito @fmassa @apaszke

### CODE REVIEW

**Review by zdevito — APPROVED** [review]
Looks good!

**Comment by nliz77** [comment]
Thx

File: test/cpp_extensions/doubler.h [hunk 1]
```diff
@@ -7,6 +7,9 @@ struct Doubler {
   at::Tensor forward() {
     return tensor_ * 2;
   }
+  at::Tensor get() const {
+    return tensor_;
+  }
```
Comment (nliz77) [inline test/cpp_extensions/doubler.h]: Thx

**Pull request merged**